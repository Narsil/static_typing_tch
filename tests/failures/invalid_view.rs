use static_typing_tch::tensor_check_fn;
use tch::Tensor;

#[tensor_check_fn]
fn transformer_mlp(
    hidden_states: &Tensor<(B, S, H)>,
    dense: &Tensor<(H, H4)>,
    dense_bias: &Tensor<(U1, H4)>,
) -> Tensor<(B, S, H)> {
    let size = hidden_states.size();
    let hidden_states = hidden_states.view((-1, size[1])); // Error here
    let hidden_states = dense_bias.addmm(&hidden_states, &dense);
    let hidden_states = hidden_states.view((size[0], size[1], size[2]));
    hidden_states
}

#[tensor_check_fn]
fn transformer_mlp2(
    hidden_states: &Tensor<(B, S, H)>,
    dense: &Tensor<(H, H4)>,
    dense_bias: &Tensor<(U1, H4)>,
) -> Tensor<(B, S, H)> {
    let size = hidden_states.size();
    let hidden_states = hidden_states.view((size[2], -1)); // Error here
    let hidden_states = dense_bias.addmm(&hidden_states, &dense);
    let hidden_states = hidden_states.view((size[0], size[1], size[2]));
    hidden_states
}

#[tensor_check_fn]
fn transformer_mlp3(
    hidden_states: &Tensor<(B, S, H)>,
    dense: &Tensor<(H, H4)>,
    dense_bias: &Tensor<(U1, H4)>,
) -> Tensor<(B, S, H)> {
    let size = hidden_states.size();
    let hidden_states = hidden_states.view((-1, size[2]));
    let hidden_states = dense_bias.addmm(&hidden_states, &dense);
    let hidden_states = hidden_states.view((size[0], size[1])); // Error here missing 1 dim
    hidden_states
}

fn main() {}
