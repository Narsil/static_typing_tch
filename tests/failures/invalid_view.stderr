error: Second dim was expected to be Symbol("H") but found Symbol("S")
  --> tests/failures/invalid_view.rs:12:42
   |
12 |     let hidden_states = dense_bias.addmm(&hidden_states, &dense);
   |                                          ^^^^^^^^^^^^^^

error: This view [Symbol("B"), Symbol("S"), Symbol("H")] is incompatible with tensor shape [Mul([Symbol("B"), Symbol("H")]), Symbol("H4")]
  --> tests/failures/invalid_view.rs:13:44
   |
13 |     let hidden_states = hidden_states.view((size[0], size[1], size[2]));
   |                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: Second dim was expected to be Symbol("H") but found Mul([Symbol("B"), Symbol("S")])
  --> tests/failures/invalid_view.rs:24:42
   |
24 |     let hidden_states = dense_bias.addmm(&hidden_states, &dense);
   |                                          ^^^^^^^^^^^^^^

error: This view [Symbol("B"), Symbol("S"), Symbol("H")] is incompatible with tensor shape [Symbol("H"), Symbol("H4")]
  --> tests/failures/invalid_view.rs:25:44
   |
25 |     let hidden_states = hidden_states.view((size[0], size[1], size[2]));
   |                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: This view [Symbol("B"), Symbol("S")] is incompatible with tensor shape [Mul([Symbol("B"), Symbol("S")]), Symbol("H4")]
  --> tests/failures/invalid_view.rs:37:44
   |
37 |     let hidden_states = hidden_states.view((size[0], size[1])); // Error here missing 1 dim
   |                                            ^^^^^^^^^^^^^^^^^^
